The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11424643025/job/31785216840) [(plain)](https://github.com/rust-lang/rust/commit/37cc13c350f1955fc7f50d5fd4c7a59ca26d8197/checks/31785216840/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
diff of stderr:

78   --> $DIR/issue-109425.rs:18:5
79    |
80 LL |     is((), 1, "", ());
-    |     ^^ --         -- unexpected argument of type `()`
+    |     ^^ --         -- unexpected argument #4 of type `()`
-    |        unexpected argument of type `()`
+    |        unexpected argument #1 of type `()`
84    |
85 note: function defined here
85 note: function defined here
86   --> $DIR/issue-109425.rs:5:4

97   --> $DIR/issue-109425.rs:20:5
98    |
99 LL |     is(1, (), "", ());
-    |     ^^    --      -- unexpected argument of type `()`
+    |     ^^    --      -- unexpected argument #4 of type `()`
-    |           unexpected argument of type `()`
+    |           unexpected argument #2 of type `()`
103    |
104 note: function defined here
---
To only update this specific test, also pass `--test-args argument-suggestions/issue-109425.rs`

error: 1 errors occurred comparing output.
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/argument-suggestions/issue-109425.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/argument-suggestions/issue-109425" "-A" "unused" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/argument-suggestions/issue-109425/auxiliary"
--- stderr -------------------------------
error[E0061]: this function takes 0 arguments but 2 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:10:5
   |
   |
LL |     f(0, 1,);        // f()
   |     ^ -  - unexpected argument #2 of type `{integer}`
   |       unexpected argument #1 of type `{integer}`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:3:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:3:4
   |
LL | fn f() {}
   |    ^
help: remove the extra arguments
   |
LL -     f(0, 1,);        // f()
LL +     f();        // f()

error[E0061]: this function takes 1 argument but 3 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:12:5
   |
   |
LL |     i(0, 1, 2,);     // i(0,)
   |     ^    -  - unexpected argument #3 of type `{integer}`
   |          unexpected argument #2 of type `{integer}`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:4:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:4:4
   |
LL | fn i(_: u32) {}
   |    ^ ------
   |
   |
LL -     i(0, 1, 2,);     // i(0,)
LL +     i(0,);     // i(0,)

error[E0061]: this function takes 1 argument but 3 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:14:5
   |
   |
LL |     i(0, 1, 2);      // i(0)
   |     ^    -  - unexpected argument #3 of type `{integer}`
   |          unexpected argument #2 of type `{integer}`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:4:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:4:4
   |
LL | fn i(_: u32) {}
   |    ^ ------
   |
   |
LL -     i(0, 1, 2);      // i(0)
LL +     i(0);      // i(0)

error[E0061]: this function takes 2 arguments but 4 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:16:5
   |
   |
LL |     is(0, 1, 2, ""); // is(0, "")
   |     ^^    -  - unexpected argument #3 of type `{integer}`
   |           unexpected argument #2 of type `{integer}`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:5:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:5:4
   |
LL | fn is(_: u32, _: &str) {}
   |    ^^ ------  -------
   |
   |
LL -     is(0, 1, 2, ""); // is(0, "")
LL +     is(0, ""); // is(0, "")

error[E0061]: this function takes 2 arguments but 4 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:18:5
   |
   |
LL |     is((), 1, "", ());
   |     ^^ --         -- unexpected argument #4 of type `()`
   |        unexpected argument #1 of type `()`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:5:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:5:4
   |
LL | fn is(_: u32, _: &str) {}
   |    ^^ ------  -------
   |
   |
LL -     is((), 1, "", ());
LL +     is(1, "");

error[E0061]: this function takes 2 arguments but 4 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:20:5
   |
   |
LL |     is(1, (), "", ());
   |     ^^    --      -- unexpected argument #4 of type `()`
   |           unexpected argument #2 of type `()`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:5:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:5:4
   |
LL | fn is(_: u32, _: &str) {}
   |    ^^ ------  -------
   |
   |
LL -     is(1, (), "", ());
LL +     is(1, "");

error[E0061]: this function takes 1 argument but 3 arguments were supplied
##[error]  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:22:5
   |
   |
LL |     s(0, 1, "");     // s("")
   |     ^ -  - unexpected argument #2 of type `{integer}`
   |       unexpected argument #1 of type `{integer}`
   |
note: function defined here
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:6:4
  --> /checkout/tests/ui/argument-suggestions/issue-109425.rs:6:4
   |
LL | fn s(_: &str) {}
   |    ^ -------
help: remove the extra arguments
   |
LL -     s(0, 1, "");     // s("")
LL +     s("");     // s("")

error: aborting due to 7 previous errors

For more information about this error, try `rustc --explain E0061`.
```

</details>

--- a/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
+++ b/compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs
@@ -1097,6 +1097,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let mut only_extras_so_far = errors
             .peek()
             .is_some_and(|first| matches!(first, Error::Extra(arg_idx) if arg_idx.index() == 0));
+        let mut prev_extra_idx = None;
         let mut suggestions = vec![];
         while let Some(error) = errors.next() {
             only_extras_so_far &= matches!(error, Error::Extra(_));
@@ -1165,11 +1166,29 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         //     fn f() {}
                         //   - f(0, 1,)
                         //   + f()
-                        if only_extras_so_far
-                            && !errors
-                                .peek()
-                                .is_some_and(|next_error| matches!(next_error, Error::Extra(_)))
-                        {
+                        let trim_next_comma = match errors.peek() {
+                            Some(Error::Extra(provided_idx))
+                                if only_extras_so_far
+                                    && provided_idx.index() > arg_idx.index() + 1 =>
+                            // If the next Error::Extra ("next") doesn't next to current ("current"),
+                            // fn foo(_: (), _: u32) {}
+                            // - foo("current", (), 1u32, "next")
+                            // + foo((), 1u32)
+                            // If the previous error is not a `Error::Extra`, then do not trim the next comma
+                            // - foo((), "current", 42u32, "next")
+                            // + foo((), 42u32)
+                            {
+                                prev_extra_idx.map_or(true, |prev_extra_idx| {
+                                    prev_extra_idx + 1 == arg_idx.index()
+                                })
+                            }
+                            // If no error left, we need to delete the next comma
+                            None if only_extras_so_far => true,
+                            // Not sure if other error type need to be handled as well
+                            _ => false,
+                        };
+
+                        if trim_next_comma {
                             let next = provided_arg_tys
                                 .get(arg_idx + 1)
                                 .map(|&(_, sp)| sp)
@@ -1192,6 +1211,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                             SuggestionText::Remove(_) => SuggestionText::Remove(true),
                             _ => SuggestionText::DidYouMean,
                         };
+                        prev_extra_idx = Some(arg_idx.index())
                     }
                 }
                 Error::Missing(expected_idx) => {

--- a/tests/ui/argument-suggestions/issue-109425.fixed
+++ b/tests/ui/argument-suggestions/issue-109425.fixed
@@ -15,6 +15,10 @@ fn main() {
     //~^ error: this function takes 1 argument but 3 arguments were supplied
     is(0, ""); // is(0, "")
     //~^ error: this function takes 2 arguments but 4 arguments were supplied
+    is(1, "");
+    //~^ error: this function takes 2 arguments but 4 arguments were supplied
+    is(1, "");
+    //~^ error: this function takes 2 arguments but 4 arguments were supplied
     s("");     // s("")
     //~^ error: this function takes 1 argument but 3 arguments were supplied
 }

--- a/tests/ui/argument-suggestions/issue-109425.rs
+++ b/tests/ui/argument-suggestions/issue-109425.rs
@@ -15,6 +15,10 @@ fn main() {
     //~^ error: this function takes 1 argument but 3 arguments were supplied
     is(0, 1, 2, ""); // is(0, "")
     //~^ error: this function takes 2 arguments but 4 arguments were supplied
+    is((), 1, "", ());
+    //~^ error: this function takes 2 arguments but 4 arguments were supplied
+    is(1, (), "", ());
+    //~^ error: this function takes 2 arguments but 4 arguments were supplied
     s(0, 1, "");     // s("")
     //~^ error: this function takes 1 argument but 3 arguments were supplied
 }

--- a/tests/ui/argument-suggestions/issue-109425.stderr
+++ b/tests/ui/argument-suggestions/issue-109425.stderr
@@ -74,8 +74,46 @@ LL -     is(0, 1, 2, ""); // is(0, "")
 LL +     is(0, ""); // is(0, "")
    |
 
-error[E0061]: this function takes 1 argument but 3 arguments were supplied
+error[E0061]: this function takes 2 arguments but 4 arguments were supplied
   --> $DIR/issue-109425.rs:18:5
+   |
+LL |     is((), 1, "", ());
+   |     ^^ --         -- unexpected argument #4 of type `()`
+   |        |
+   |        unexpected argument #1 of type `()`
+   |
+note: function defined here
+  --> $DIR/issue-109425.rs:5:4
+   |
+LL | fn is(_: u32, _: &str) {}
+   |    ^^ ------  -------
+help: remove the extra arguments
+   |
+LL -     is((), 1, "", ());
+LL +     is(1, "");
+   |
+
+error[E0061]: this function takes 2 arguments but 4 arguments were supplied
+  --> $DIR/issue-109425.rs:20:5
+   |
+LL |     is(1, (), "", ());
+   |     ^^    --      -- unexpected argument #4 of type `()`
+   |           |
+   |           unexpected argument #2 of type `()`
+   |
+note: function defined here
+  --> $DIR/issue-109425.rs:5:4
+   |
+LL | fn is(_: u32, _: &str) {}
+   |    ^^ ------  -------
+help: remove the extra arguments
+   |
+LL -     is(1, (), "", ());
+LL +     is(1, "");
+   |
+
+error[E0061]: this function takes 1 argument but 3 arguments were supplied
+  --> $DIR/issue-109425.rs:22:5
    |
 LL |     s(0, 1, "");     // s("")
    |     ^ -  - unexpected argument #2 of type `{integer}`
@@ -93,6 +131,6 @@ LL -     s(0, 1, "");     // s("")
 LL +     s("");     // s("")
    |
 
-error: aborting due to 5 previous errors
+error: aborting due to 7 previous errors
 
 For more information about this error, try `rustc --explain E0061`.

--- a/tests/ui/argument-suggestions/issue-112507.stderr
+++ b/tests/ui/argument-suggestions/issue-112507.stderr
@@ -18,8 +18,9 @@ LL |     Float(Option<f64>),
    |     ^^^^^
 help: remove the extra arguments
    |
-LL ~         ,
-LL ~         None);
+LL -         0,
+LL -         None,
+LL +         None);
    |
 
 error: aborting due to 1 previous error

